rec: Allow multiple simultaneous incoming TCP queries over a connection

omoerbeek opened the pull request:
Fixes #8358 

### Short description
<!-- Write a small description of what this Pull Request fixes or provides, including the issue #s -->

Answers are sent out the moment the become available, so not necessarily in the
same order as received. There's a limit on how many queries per TCP
induced connection we may have in flight.

Only very lightly tested!

Includes an sdig hack to read multiple queries from a file.

### Checklist
<!-- please indicate if any of these things are done/included with this Pull Request. Not all boxes need to be checked for the Pull Request to be accepted -->
I have:
- [X] read the [CONTRIBUTING.md](https://github.com/PowerDNS/pdns/blob/master/CONTRIBUTING.md) document
- [X] compiled this code
- [X] tested this code
- [x] included documentation (including possible behaviour changes)
- [ ] documented the code
- [x] added or modified regression test(s)
- [ ] added or modified unit test(s)

rgacogne commented on pdns/syncres.hh:
```diff
@@ -1012,8 +1012,10 @@ public:
   uint16_t qlen{0};
   uint16_t bytesread{0};
-
+  std::atomic<int> d_requestsInFlight;
```
Does it really need to be atomic? I don't think we have any concurrency issue?

omoerbeek replied:
I wasn't sure if tcpconnection was read from multiple threads. Will double check.

omoerbeek replied:
Indeed, TCPConnections are only shared by mthreads

rgacogne commented on pdns/syncres.hh:
```diff
@@ -1012,8 +1012,10 @@ public:
   uint16_t qlen{0};
   uint16_t bytesread{0};
-
+  std::atomic<int> d_requestsInFlight;
```
Also just a nit but I'd prefer an unsigned type of known size :-)

rgacogne commented on pdns/pdns_recursor.cc:
```diff
@@ -2002,7 +2008,11 @@ static void handleRunningTCPQuestion(int fd, FDMultiplexer::funcparam_t& var)
+      if (conn->d_requestsInFlight >= TCPConnection::s_maxInFlight) {
+        //cerr << "Disabling... " << conn->d_requestsInFlight << ' ' << fd << endl;
+        t_fdm->removeReadFD(fd); // should no longer awake ourselves when there is data to read
+      }
```
Shouldn't we update the read TTD when we don't disable the polling?

omoerbeek replied:
yes, I think so. The original code didn't do that, but it should have.

rgacogne commented on pdns/pdns_recursor.cc:
```diff
@@ -1760,12 +1762,16 @@ static void startDoResolve(void *p)
+            t_fdm->addReadFD(dc->d_socket, handleRunningTCPQuestion, dc->d_tcpConnection, &ttd);
+          } else {
+            t_fdm->setReadTTD(dc->d_socket, ttd,  g_tcpTimeout);
+          }
```
It looks like we disable the polling on a timeout even if we have queries in-flight (good), but we could resume it later if `dc->d_tcpConnection->d_requestsInFlight == TCPConnection::s_maxInFlight - 1` which might not be what we want?

rgacogne commented on pdns/pdns_recursor.cc:
```diff
@@ -1760,12 +1762,16 @@ static void startDoResolve(void *p)
+            ttd.tv_sec += g_tcpTimeout;
+            t_fdm->addReadFD(dc->d_socket, handleRunningTCPQuestion, dc->d_tcpConnection, &ttd);
+          } else {
+            t_fdm->setReadTTD(dc->d_socket, ttd,  g_tcpTimeout);
```
Would that fail if the polling was disabled because of a timeout but we still had at least one query in-flight?

omoerbeek replied:
Will look into these timeout related issues tomorrow.

omoerbeek replied:
Right in flight count should be decremented always and the fd should be remove if it is in the set

omoerbeek commented:
I think I addressed the FD accounting. It's a bit more complex than you'd think, because you don't know if another mthread before you already had a write error and removed the fd from the multiplexer.

rgacogne commented on pdns/pdns_recursor.cc:
```diff
@@ -1748,24 +1750,50 @@ static void startDoResolve(void *p)
+          if (dc->d_tcpConnection->d_requestsInFlight == TCPConnection::s_maxInFlight - 1) {
+            // A read error might have happened. If we add the fd back, it will most likely error again.
+            // This is not a big issue, the next handleTCPClientReadable() will see another read error
+            // and take action.
```
One possible issue here is that we might have decided to stop reading because the idle timeout was reached, and in that case we will start reading again. I wonder if that might open us to some sort of slow loris attack where a client can keep the TCP connection alive for a very long time without reaching the maximum number of queries on a single TCP connection.

omoerbeek commented:
The code in the `GETQUESTION` block case needs some fixes, since there are multiple exit paths. In those case the state isn't updated and the `d_requestsInFlight` counter will be wrong.

Additionally, a regression test would be nice.

As for the slow loris attack, I'll try to see if this is possible and/or how to defend against it.

omoerbeek commented:
As for the slow loris attack: the FD is only added back after we processed a query. If we want to come back in the same state again, we can only do that by processing another query, AFAIKS. So the query count does get increased.

omoerbeek commented:
Some remaining questions to be answered:
- do we want the sdig part (it includes some refactoring that might be nice)
- what should be the default for `max-concurrent-requests-per-tcp-connection`? Currently 10, but we might be conservative and make it 1.

rgacogne commented on regression-tests.recursor-dnssec/test_SimpleTCP.py:
```diff
@@ -0,0 +1,123 @@
+        self.assertRRsetInAnswer(res, expectedA)
+
+
+    def testVeryBasicOOO(self):
```
Would that test fail without OOO support? It looks like it would work as long as we support pipelining.

omoerbeek replied:
I added a note in the test file I'm not sure yet how to test the OOO code actually.

rgacogne commented on pdns/syncres.hh:
```diff
@@ -1012,7 +1012,9 @@ public:
-
+  uint16_t d_requestsInFlight{0}; // number of mthreads spawned for this connection
+  // The max number of concurrent TCP requests we're willing to process
+  uint16_t d_maxInFlight;
```
I'm wondering if that could not be a static member, only updated once instead of reading `max-concurrent-requests-per-tcp-connection` for each incoming TCP connection?

omoerbeek commented:
Actually, looking back, what I implemented is pipelining with the additionall bonus of ooo processing.

For unbound it was implemented in Feb 2019, so pretty recent.  For knot and bind resolver IO could not find details, except that they do support pipelining/ooo according to https://dnsprivacy.org/wiki/display/DP/DNS+Privacy+Implementation+Status 

I'll look into making the test a bit more extensive and the static thing.

cmouse commented on pdns/pdns_recursor.cc:
```diff
@@ -1748,24 +1749,49 @@ static void startDoResolve(void *p)
-      // update tcp connection status, either by closing or moving to 'BYTE0'
+      // update tcp connection status, closing if needed and doing the fd multiplexer accounting
 
+      dc->d_tcpConnection->d_requestsInFlight--;
```
could be a good idea to make sure d_requestsInFlight>0 here...

omoerbeek commented:
As for the default value, I think I'll leave it at 10 to see if alpha tester catch anything.

cmouse commented on pdns/pdns_recursor.cc:
```diff
@@ -1778,7 +1780,7 @@ static void startDoResolve(void *p)
           Utility::gettimeofday(&g_now, 0); // needs to be updated
           struct timeval ttd = g_now;
-          if (dc->d_tcpConnection->d_requestsInFlight == dc->d_tcpConnection->d_maxInFlight - 1) {
+          if (dc->d_tcpConnection->d_requestsInFlight == TCPConnection::s_maxInFlight - 1) {
```
this would be better formed as `dc->d_tcpConnection->d_requestsInFlight + 1 == TCPConnection::s_maxInFlight` given it's unsigned int.

omoerbeek replied:
`TCPConnection::s_maxInFlight` should be `>= 1` (though we don't check for that atm).

omoerbeek commented:
BTW tests after latest commit will probably fail because it needs the lua sockets lib.

Habbie commented:
> BTW tests after latest commit will probably fail because it needs the lua sockets lib.

Fix for that being tested in https://github.com/Habbie/pdns/tree/rec-out-of-order-luasocket - if that goes green on circle and travis, you can steal that commit.

Habbie commented:
> Fix for that being tested in https://github.com/Habbie/pdns/tree/rec-out-of-order-luasocket - if that goes green on circle and travis, you can steal that commit.

This went green (except for Travis auth which is unrelated) so I pushed the commit here.

pieterlexis commented on pdns/pdns_recursor.cc:
```diff
@@ -4014,6 +4014,13 @@ static int serviceMain(int argc, char*argv[])
 
+  
+  if (TCPConnection::s_maxInFlight < 1 || TCPConnection::s_maxInFlight > USHRT_MAX || TCPConnection::s_maxInFlight >= g_maxMThreads) {
+    g_log<<Logger::Warning<<"Asked to run with illegal max-concurrent-requests-per-tcp-connection, setting to default"<<endl;
```
You might want to log the default as well :)

omoerbeek replied:
Oh, how I wish the default could be accessed by code ....

omoerbeek replied:
Actually, the logic is wrong. I need to validate before assigning to a ushort.
